vulcanai2/tests/models/test_cnn.py

```diff
@@ -1,114 +1,204 @@
-        )
+    def multi_input_cnn_train_loader(self, multi_input_cnn_data):
+        """Synthetic test data pytorch dataloader object."""
+        test_train = Subset(multi_input_cnn_data,
+                            range(len(multi_input_cnn_data)//2))
+        return DataLoader(test_train, batch_size=2)
 
     @pytest.fixture
-    def cnn_class(self):
-        """Create ConvNet with prediction layer."""
-        return ConvNet(
-            name='Test_ConvNet_class',
-            in_dim=(1, 28, 28),
-            config={
-                'conv_units': [
-                    {
-                        "in_channels": 1,
-                        "out_channels": 16,
-                        "kernel_size": (5, 5),
-                        "pool_size": 2,
-                        "stride": 2
-                    },
-                    {
-                        "in_channels": 16,
-                        "out_channels": 1,
-                        "kernel_size": (5, 5),
-                        "stride": 2,
-                        "padding": 2
-                    }]
-            },
-            num_classes=3
-        )
-
-    def test_forward_not_nan(self, cnn_noclass):
+    def multi_input_cnn_test_loader(self, multi_input_cnn_data):
+        """Synthetic train data pytorch dataloader object."""
+        test_val = Subset(multi_input_cnn_data,
+                          range(len(multi_input_cnn_data)//2,
+                                len(multi_input_cnn_data)))
+        return DataLoader(test_val, batch_size=2)
+
+    def test_init(self, conv1D_net):
+        """Initialization Test of a ConvNet object."""
+        assert isinstance(conv1D_net, BaseNetwork)
+        assert isinstance(conv1D_net, nn.Module)
+        assert hasattr(conv1D_net, 'network')
+        assert hasattr(conv1D_net, 'in_dim')
+        assert hasattr(conv1D_net, 'record')
+        assert hasattr(conv1D_net, 'device')
+
+        assert conv1D_net._name is not None
+        assert isinstance(conv1D_net._config, ConvNetConfig)
+
+        assert conv1D_net.input_networks is None
+        assert conv1D_net.epoch == 0
+        assert conv1D_net.optim is None
+        assert conv1D_net.criterion is None
+
+        assert not hasattr(conv1D_net, 'metrics')
+
+    def test_function_multi_input(self, conv1D_net, multi_input_cnn):
+        """Test functions wrt multi_input_cnn."""
+        assert isinstance(multi_input_cnn.input_networks, nn.ModuleDict)
+        assert len(list(multi_input_cnn.input_networks)) == 3
+        assert all(multi_input_cnn._get_max_incoming_spatial_dims() == (8, 8, 8))
+        assert multi_input_cnn._merge_input_network_outputs([
+                torch.ones([10, 1, 28, 28]),
+                torch.ones([10, 1, 28, 28, 28]),
+                torch.ones(10, *multi_input_cnn.
+                           input_networks['multi_input_dnn'].out_dim)
+            ]).shape == (10, 3, 8, 8, 8)
+        test_net = pickle.loads(pickle.dumps(multi_input_cnn))
```

Comment:
?? Can someone explain why this is here?

Reply:
why this instead of deepcopy? I'm pretty sure it would be faster.

Reply:
good to know. didn't know deepcopy is faster. I just tested it. Thanks.. I changed it to deepcopy now